File: medium/155-min-stack.ts
// 

interface Stack {
  val: number,
  min: number
}

class MinStack {
  public stack: Stack[];
  
  constructor() {
    this.stack = [];
  }

  push(val: number): void {
    this.stack.push({
      val,
      min: this.stack.length == 0 ? val: Math.min(val, this.getMin())
    });
  }

  pop(): void {
    this.stack.pop();
  }

  top(): number {
    return this.stack[this.stack.length - 1].val;
  }

  getMin(): number {
    return this.stack.length == 0 ? 0: this.stack[this.stack.length - 1].min;
  }
}

const test = new MinStack();
test.push(-2);
test.push(0);
test.push(-3);
console.log(test.getMin());
test.pop();
console.log(test.top());
console.log(test.getMin());

export default MinStack;

File: python/155-min-stack.py
class MinStack:
    def __init__(self):
        self.cache = []
        self.minCache = []

    def push(self, val: int) -> None:
        self.cache.append(val)
        val = min(val, self.minCache[-1] if self.minCache else val)
        self.minCache.append(val)

    def pop(self) -> None:
        self.minCache.pop()
        return self.cache.pop()

    def top(self) -> int:
        return self.cache[-1]

    def getMin(self) -> int:
        return self.minCache[-1]